### DISCUSSION TITLE
Triply periodic boundary conditions

### DISCUSSION THREAD

**Discussion by makrandak** [Nek5000/nekRS#556] [Q&A]
Hello, 

I am trying to prescribe periodic boundary conditions for all pairs of faces on a box (periodic in x, y, z). I generated the box using genbox. Below is the script for the generation of the box
```
base.rea
-3                     spatial dimension  ( < 0 --> generate .rea/.re2 pair)
1                      number of fields
#=======================================================================
#
#    Example of .box file for Taylor-Green
#
#    If nelx (y or z) < 0, then genbox automatically generates the
#                          grid spacing in the x (y or z) direction
#                          with a geometric ratio given by "ratio".
#                          ( ratio=1 implies uniform spacing )
#
#    Note that the character bcs _must_ have 3 spaces.
#
#=======================================================================
#
Box
-16  -16  -16                                        nelx,nely,nelz for Box
-1.570796327 1.570796327 1.                          x0,x1,gain  (rescaled in usrdat)
-1.570796327 1.570796327 1.                          y0,y1,gain  (rescaled in usrdat)
-1.570796327 1.570796327 1.                          z0,z1,gain
P  ,P  ,P  ,P  ,P  ,P                                bc's  (3 chars each!)
````

I also prescribe the following in the `.usr` file to make sure the boundaryIDs are zero for all faces (following the documentation), I skipped the exo2nek step because I generated the mesh using genbox. 

usrdat2 is as follows. 

```
c-----------------------------------------------------------------------
      subroutine usrdat2()  ! This routine to modify mesh coordinates
      include 'SIZE'
      include 'TOTAL'
      
      integer e,f

      n = lx1*ly1*lz1*nelv
      nxz = nx1*nz1
      nface = 2*ldim

      do iel=1,nelt
         cbc(1,iel,1) = 'P  '
         cbc(2,iel,1) = 'P  '
         cbc(3,iel,1) = 'P  '
         cbc(4,iel,1) = 'P  '
         cbc(5,iel,1) = 'P  '
         cbc(6,iel,1) = 'P  '
      do ifc=1,2*ndim
         cbc(ifc,iel,2)  = cbc(ifc,iel,1) 
         if (cbc(ifc,iel,1) .eq. 'P  ') boundaryID(ifc,iel) = 0
         if (cbc(ifc,iel,2) .eq. 'P  ') boundaryID(ifc,iel) = 0
         if (cbc(ifc,iel,3) .eq. 'P  ') boundaryID(ifc,iel) = 0
         if (cbc(ifc,iel,4) .eq. 'P  ') boundaryID(ifc,iel) = 0
         if (cbc(ifc,iel,5) .eq. 'P  ') boundaryID(ifc,iel) = 0
         if (cbc(ifc,iel,6) .eq. 'P  ') boundaryID(ifc,iel) = 0
      enddo
      enddo

      return
      end
c-----------------------------------------------------------------------
```

 When I try to run a Taylor-Green vortex case with the setup above, I see non-periodic behaviour for x-velocity in x direction, y-velocity in y direction and so on.  Is there something else I have to do to prescribe periodicity? 
 Also, I found some this routine in the ktauChannel setup saying kludge for periodic in z. I am not sure if this is required for periodicity
 ```
 subroutine usrsetvert(glo_num,nel,nx,ny,nz) ! to modify glo_num
      integer*8 glo_num(1)

      ! kludge for periodic bc in z
      nxy  = nx*ny
      nxyz = nx*ny*nz
      do iel = 1,nel
         ioff = nxyz*(iel-1)
         do ixy = 1,nxy
            glo_num(ioff + nxy*(nz-1) + ixy) = glo_num(ioff + ixy)
         enddo
      enddo

      return
      end
```
I am new to nekRS and I have not used nek5000 in the past. I am not sure what I am doing wrong in the setup. I would really appreciate your help.  I would be happy to share all case files if needed to reproduce the problem. 

With sincere regards

**Comment by yslan** [comment]
Hi, 

Can you check `ldimt` in .cache/nek5000/SIZE ?
cbc is of size `cbc(6,lelt,0:ldimt1)` I think 6 (from `(cbc(ifc,iel,6)`) is uncommonly large and you could access the memory for other variables.

You don't need to add `usersetvert` since the periodicity should be directly coming from ma2/parCON. 
In fact, I don't think you need to set boundaryID at all. NekRS should read cbc as-is.

I suggest to comment all those out.
Or, it should at least work if you simply replace the mesh in the tgv example.

Hope this helps,
Yu-Hsiang

**Comment by makrandak** [comment]
Hello Yu-Hsiang,

I commented all of it out and added boundaryTypeMap = none​ in my .par​ file. My box file is pretty much the same as the one in tgv example except for fewer elements (domain is rescled in `.usr` file in the tgv) example. I still get non-periodic results.

How is the mesh in the tgv example generated? from the box file?

With sincere regards,
Makrand


Makrand A. Khanwale
-------------------------------------------
Postdoctoral Fellow,
Center for Turbulence Research,
Department of Mechanical Engineering,
Stanford University​
​Website: makrandak.github.io<http://makrandak.github.io/>
Email: ***@***.******@***.***>
________________________________

**Reply by yslan** [reply]
Yes, there is a box file here
https://github.com/Nek5000/nekRS/blob/master/examples/tgv/input.box

Your box file looks fine though. It should work.
I recommend to rescale the domain in usrdat2 since box file only deal with ascii with a fewer digits.

I wonder if you are indeed using the re2 from genbox.
The output from genbox should be a `box.re2` file and you typically need to rename is to match your <casename>.par

**Comment by makrandak** [comment]
Hello Yu-Hsiang,

I am doing exactly that. I am using genbox and then renaming the file to match the .par name. I can see the output in visit and the mesh matches my settings in the box file.

I have commented all of that stuff from the usr file and the udf file is something I wrote.  The original effort was to setup a periodic HIT simulation with linear forcing, but I wanted to try Taylor-Green first as I was getting discontinuous solutions when I prescribed the energy spectrum based initial conditions.

Right now the only thing in udf that is interesting is the initial condition, nothing to do with boundary conditions.

I am attaching my case files to the email.

Thanks again for your help.

With sincere regards,
Makrand


Makrand A. Khanwale
-------------------------------------------
Postdoctoral Fellow,
Center for Turbulence Research,
Department of Mechanical Engineering,
Stanford University​
​Website: makrandak.github.io<http://makrandak.github.io/>
Email: ***@***.******@***.***>
________________________________

**Comment by makrandak** [comment]
I think I found a bug, my domain was from -pi/2 to pi/2 instead of pi to pi.

I will test further and let you know. Thank you very much for your help and time.

With sincere regards,
Makrand


Makrand A. Khanwale
-------------------------------------------
Postdoctoral Fellow,
Center for Turbulence Research,
Department of Mechanical Engineering,
Stanford University​
​Website: makrandak.github.io<http://makrandak.github.io/>
Email: ***@***.******@***.***>
________________________________
